lib/Conversion/AffineToPipeline/AffineToPipeline.cpp

```diff
@@ -408,37 +467,103 @@ LogicalResult AffineToPipeline::createPipelinePipeline(
 
   // Iterate in order of the start times.
   SmallVector<unsigned> startTimes;
-  for (auto group : startGroups)
+  for (const auto &group : startGroups)
     startTimes.push_back(group.first);
   llvm::sort(startTimes);
 
   DominanceInfo dom(getOperation());
+
+  // Keys for translating values in each stage
+  SmallVector<SmallVector<Value>> registerValues;
+  SmallVector<SmallVector<mlir::Type>> registerTypes;
+
+  // The maps that ensure a stage uses the correct version of a value
+  SmallVector<BlockAndValueMapping> stageValueMaps;
+
+  // For storing the range of stages an operation's results need to be valid for
+  DenseMap<Operation *, std::pair<unsigned, unsigned>> pipeTimes;
+
   for (auto startTime : startTimes) {
     auto group = startGroups[startTime];
-    OpBuilder::InsertionGuard g(builder);
 
     // Collect the return types for this stage. Operations whose results are not
     // used within this stage are returned.
     auto isLoopTerminator = [forOp](Operation *op) {
       return isa<AffineYieldOp>(op) && op->getParentOp() == forOp;
     };
-    SmallVector<Type> stageTypes;
-    DenseSet<Operation *> opsWithReturns;
+
+    // Initialize set of registers up until this point in time
+    for (unsigned i = registerValues.size(); i <= startTime; ++i)
+      registerValues.push_back(SmallVector<Value>());
+
+    // Check each operation to see if its results need plumbing
     for (auto *op : group) {
+      if (op->getUsers().empty())
+        continue;
+
+      unsigned pipeStartTime = std::numeric_limits<unsigned>::max();
+      unsigned pipeEndTime = 0;
       for (auto *user : op->getUsers()) {
-        if (*problem.getStartTime(user) > startTime || isLoopTerminator(user)) {
-          opsWithReturns.insert(op);
-          stageTypes.append(op->getResultTypes().begin(),
-                            op->getResultTypes().end());
+        if (*problem.getStartTime(user) > startTime) {
+          pipeEndTime = std::max(pipeEndTime, *problem.getStartTime(user));
```

Comment:
Nit: Maybe introduce a `userStartTime` variable to replace the four calls to `getStartTime` for improved readability.